src/fges.jl

```diff
@@ -0,0 +1,368 @@
+    isblocked(g, x, y, nodesRemoved) = !has_a_path(g, [x], y, nodesRemoved)
+
+    # Calculate two (possibly empty) sets of nodes
+    # NAxy: any nodes that are undirected neighbors of y and connected to x by any edge
+    # Txy: any subset of the undirected neighbors of y not connected to x
+    Tyx = calcT(g, y, x)
+    NAyx = calcNAyx(g, y, x)
+
+
+    # Create two containers to hold the best found score and best subset of Tyx
+    bestScore = zero(step.Δscore)
+    bestT = Vector{Int}()
+
+    # Keep a list of invalid sets
+    invalid = Vector{Vector{Int}}()
+    
+    # Loop through all possible subsets of Tyx
+    for T in powerset(Tyx)
+        if checkSupersets(T,invalid)
+            NAyxT = NAyx ∪ T
+            if isclique(g, NAyxT) && isblocked(g, y, x, NAyxT)
+
+                # Score the valid Insert
+                PAy = parents(g,y)
+                PAy⁺ = PAy ∪ x
+                newScore = score(dataParsed, NAyxT ∪ PAy⁺, y) - score(dataParsed, NAyxT ∪ PAy, y)
+                
+                # Save the new score if it was better than any previous
+                if newScore > bestScore
+                    bestT = T
+                    bestScore = newScore
+                end
+            end
+        else
+            # Record that the subset T is invalid
+            push!(invalid,T)
+        end
+    end
+
+    return (bestT, bestScore)
+end
+
+# Check if the set T is contained in any invalid set
+function checkSupersets(T, invalid)
+    for i ∈ invalid
+        if i ⊆ T
+            return false
+        end
+    end
+    return true
+end
+
+
+function findBestDelete(step, dataParsed, g, x, y)
+    # Calculate two (possibly empty) sets of nodes
+    # NAxy: any nodes that are undirected neighbors of y and connected to x by any edge
+    # Hyx: any subset of the undirected neighbors of y that are connected to x
+    NAyx = calcNAyx(g, y, x)
+    Hyx = NAyx
+
+    # Ceate two containers to hold the best found score and best subset of Tyx
+    bestScore = zero(step.Δscore)
+    bestH = Vector{Int}()
+
+    # Loop through all possible subsets of Hyx
+    for H in powerset(Hyx)
+        # Calculate NAyx \ {H}
+        NAyx_H = setdiff(NAyx,H)
+
+        # Check if the operator is valid
+        if isclique(g, NAyx_H)
+
+            # Score the valid operator 
+            PAy = parents(g,y)
+            PAy⁻ = setdiff(PAy, x)
+            newScore = score(dataParsed, NAyx_H ∪ PAy⁻, y) - score(dataParsed, NAyx_H ∪ PAy, y)
+            
+            if newScore > bestScore
+                bestH = H
+                bestScore = newScore
+            end
+        end
+    end
+
+    return (bestH, bestScore)
+end
+
+
+####################################################################
+# Scoring function
+####################################################################
+
+@memoize LRU(maxsize=1_000_000) function score(dataParsed::ParseData{Matrix{A}}, nodeParents, node) where A
```

Comment:
For proper testing we really need an oracle version of this.

Reply:
I think so as well

Reply:
I am looking for a score equivalent oracle score. Score equivalent 
says that two dag with the same cpdag need to have the same score sum. I thought 
```
function score(cpdag::DiGraph, vparents, v)
    # possible parents are good (otherwise we could learn only the v structures)
    # impossible parents are bad
    return length(neighbors_undirected(cpdag, v) ∩ vparents)  -  length(setdiff(vparents, neighbors_adjacent(cpdag, v)))
end
```
The test are broken, not sure if this is the wrong oracle score or there are still bugs

Reply:
Hmm, I don't think it's that easy...

Maybe I misunderstand but e.g. for the CPDAG a - b - c, how do you penalize a -> b <- c which would introduce a new v-structure?